csharp/src/Ice/Reference.cs

```diff
@@ -584,6 +584,101 @@ static void StartQueryOption(StringBuilder sb, ref bool firstOption)
+        {
+            Connection? connection = GetCachedConnection();
+            bool cached = false;
+            IReadOnlyList<Endpoint>? endpoints = ImmutableList<Endpoint>.Empty;
+            List<Connector>? connectors = null;
+
+            if (connection == null || (!IsFixed && !connection.IsActive))
+            {
+                // No cached connection, so now check if there's a connection to one of our endpoints if the
+                // prefer existing connection property is true.
+
+                // TODO replace IsConnectionCached with PreferExistingConnection
+                if (IsConnectionCached)
+                {
+                    // Compute the endpoints (eventually performing a locator lookup, moving last the recently
+                    // failed endpoints, filtering endpoints based on the proxy properties). If the endpoints
+                    // are retrieved from a cache, a cached serial is returned and identifies the lookup of the
+                    // cached entry. This is useful to ensure we don't perform a new lookup if the cached entry
+                    // has already been updated since we got it.
+                    (cached, endpoints) = await ComputeEndpointsAsync(cancel).ConfigureAwait(false);
+
+                    // The communicator returns an existing  connection if one is already established to one of the given
+                    // endpoints.
+                    OutgoingConnectionFactory connectionFactory = Communicator.OutgoingConnectionFactory;
+                    connection = connectionFactory.GetConnection(endpoints, ConnectionId);
+
+                    // If we couldn't find the connection using the endpoints, expand the endpoints to
+                    // connectors (DNS lookup for IP endpoints) and try to obtain a connection using the
+                    // connectors.
+                    if (connection == null)
+                    {
+                        // This will throw if no connectors can be computed.
+
+                        // TODO retry with fresh endpoints if ComputeConnectorsAsync throws and we are using cached
+                        // endpoints.
+                        connectors = await connectionFactory.ComputeConnectorsAsync(
+                            this,
+                            endpoints,
+                            ImmutableList<Connector>.Empty,
+                            cancel).ConfigureAwait(false);
+
+                        // Lookup the connection using the connectors.
+                        connection = connectionFactory.GetConnection(connectors, ConnectionId);
+                    }
+                    _connection = connection;
+                }
+            }
+            return (connection, cached, endpoints, connectors);
+        }
+
         internal static Task<IncomingResponseFrame> InvokeAsync(
```

Comment:
The readability of this method isn't improving, we need to look into fixing this :smile:

Could we remove the inner PerformInvokeAsync method and move the code to a new Reference.InvokeAsync non-static method instead and instead of calling `return PerformInvokeAsync(request, oneway, progress, cancel);` we could call:
```
                    // After we went down the interceptor chain make the invocation.
                    request.Finish();
                    Reference reference = proxy.IceReference;

                    IInvocationObserver? observer = ObserverHelper.GetInvocationObserver(proxy,
                                                                                         request.Operation,
                                                                                         request.Context);

                    // If the request size is greater than Ice.RetryRequestSizeMax or the size of the request
                    // would increase the buffer retry size beyond Ice.RetryBufferSizeMax we release the request
                    // after it was sent to avoid holding too much memory and we wont retry in case of a failure.
                    int requestSize = request.Size;
                    bool releaseRequestAfterSent =
                        requestSize > reference.Communicator.RetryRequestSizeMax ||
                        !reference.Communicator.IncRetryBufferSize(requestSize);

                    try
                    {
                        reference.InvokeAsync(request, oneway, progress, requestRequestAfterSent, observer, cancel);
                    }
                    finally
                    {
                        if (!releaseRequestAfterSent)
                        {
                            reference.Communicator.DecRetryBufferSize(requestSize);
                        }
                        // TODO release the request memory if not already done after sent.
                        // TODO: Use IDisposable for observers, this will allow using "using".
                        observer?.Detach();
                    }
```

And we could move back the static InvokeAsync method to the proxy class (we keep the inner InvokeWithInterceptorsAsync method in the proxy InvokeAsync method).

Reply:
updated the code to use a private non-static PerformInvokeAsync